### DISCUSSION TITLE
Decoding with ARPA LM

### DISCUSSION THREAD

**Discussion by matthewkperez** [speechbrain/speechbrain#1848] [Q&A]
Hello,
I've currently finetuned a wav2vec2 model and I'd like to incorporate a N-gram LM for decoding rather than using greedyCTC decoding.

I've seen some posts which use pyctcdecode but my current tokenizer is CTCTextEncoder (i.e. just characters) and the examples which I've seen (https://github.com/speechbrain/speechbrain/issues/1467), use a unigram tokenizer.

Does anyone have suggestions on how to load a .arpa file for n-gram decoding? and how to incorporate this into sb.Brain.compute_objectives()?

**Comment by helleuch** [comment]
Hello,
I am also trying to use an n-gram model in the decoding, although for a different use case (ASR). Were you able to implement it ?

**Comment by matthewkperez** [comment]
Yes, here was my implementation

`

# main function (init pyctcdecoder and establish it as a object in asr_brain)

sorted_tokenizer = sorted(asr_brain.tokenizer.lab2ind.items(), key=lambda x:x[1])
labels = [lab for lab,ind in sorted_tokenizer]


labels[0] = '' #replace <blank> with '' to make it pyctcdecode friendly

arpa_path="/y/mkperez/AphasiaBank/boba_AB/kd_updated_para/lang-3-gram_lower/G.arpa"
asr_brain.ngram_decoder = build_ctcdecoder(labels,
    arpa_path,
    alpha=0.6)




# In compute_objectives

dec_out = self.ngram_decoder.decode(p_ctc[0].detach().cpu().numpy())
predicted_words = [dec_out.split()]
`

**Reply by helleuch** [reply]
Thank you very much !

**Reply by helleuch** [reply]
> dec_out = self.ngram_decoder.decode(p_ctc[0].detach().cpu().numpy())
> predicted_words = [dec_out.split()

What is the p_ctc[0] parameter ? Is it the obtained hypotheses by the asr_brain ?

**Reply by matthewkperez** [reply]
Is the softmax probabilities from the CTC layer. Not the hypotheses. Also, this is for a batch_size of 1 for inference decoding so that why I take the 0 index as this return a (N,P) matrix where N is the length of the sequence and P is the number of characters.

**Reply by helleuch** [reply]
Thank you very much for your help !

**Comment by Adel-Moumen** [comment]
Hi, 

recently we introduced an example of PyCTCDecode in Speechbrain, see: https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py#L108 and https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/CTC/hparams/downsampled/train_hf_wavlm_signal_downsampling.yaml 

PyCTCDecode allows N-gram decoding. 

Regarding CTCTextEncoder, the yaml example provided is using CTCTextEncoder :-)

**Reply by helleuch** [reply]
Thank you, I will have a look :)
